docs/proposals/api-group.md

```diff
@@ -0,0 +1,104 @@
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Supporting multiple API groups
+
+## Goal
+
+1. Breaking the monolithic v1 API to modular groups
+
+2. Supporting different versions in different groups
+
+3. Supporting identically named kinds to exist in different groups
+
+4. Enabling/disabling groups individually
+
+5. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
+
+6. Allowing users to omit group name when using kubectl
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  We will not add new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified, the function should search groups in a predetermined order.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+    * for api/v1, we set the prefix="api" (which is populated from app.APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
+
+    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
+
+    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
+
+2. handling common API objects like api.Status:
+
+  To handle the API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
+
+  Currently, for api.Status, we choose the latter approach: the server will encode it if the codec is v1, otherwise the server will send it unversioned.
+
+## Client-side implementation
+
+1. Client struct:
+
+  Currently each group has a client struct wrapping over \*RESTClient, e.g., client/unversioned.ExperimentalClient, client/unversioned.Client. The reason we have separate client struct is because each of them also implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). We don't know yet how to make their implementation scalable.
```

Comment:
That seems very likely - the proposal I've been considering would be that
/kube-api would list groups and version, and contain links that are
relative or absolute to where those group/version API docs are (/oapi/v1/),
and possibly their swagger docs.  That would require a way to register a
3rd party endpoint.  It also requires us to understand how to auth to those
servers in the long run, because we can't assume they have the same
authorization.

On Thu, Aug 20, 2015 at 2:29 PM, Chao Xu <notifications@github.com> wrote:

> In docs/proposals/api-group.md
> <https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37564322>
> :
>
> > +
> > +    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
> > +
> > +    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
> > +
> > +2. handling common API objects like api.Status:
> > +
> > +  To handle the API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
> > +
> > +  Currently, for api.Status, we choose the latter approach: the server will encode it if the codec is v1, otherwise the server will send it unversioned.
> > +
> > +## Client-side implementation
> > +
> > +1. Client struct:
> > +
> > +  Currently each group has a client struct wrapping over \*RESTClient, e.g., client/unversioned.ExperimentalClient, client/unversioned.Client. The reason we have separate client struct is because each of them also implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). We don't know yet how to make their implementation scalable.
>
> @smarterclayton <https://github.com/smarterclayton>, for the discovery,
> how about we let the server expose a path /kube-api, which will return a
> list of supported groups and versions in the form of "group/version"? Does
> that solve the discovery issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12951/files#r37564322>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift